src/gluonts/torch/model/deepar/deepar_network.py

```diff
@@ -0,0 +1,430 @@
+from gluonts.torch.modules.distribution_output import (
+    DistributionOutput,
+    StudentTOutput,
+)
+from gluonts.torch.modules.scaler import MeanScaler, NOPScaler
+from gluonts.torch.modules.feature import FeatureEmbedder
+from gluonts.torch.util import weighted_average
+
+
+class DeepARNetwork(nn.Module):
+    @validated()
+    def __init__(
+        self,
+        freq: str,
+        prediction_length: int,
+        num_feat_dynamic_real: int,
+        num_feat_static_real: int,
+        num_feat_static_cat: int,
+        cardinality: List[int],
+        embedding_dimension: Optional[List[int]] = None,
+        num_layers: int = 2,
+        num_cells: int = 40,
+        cell_type: str = "LSTM",
+        context_length: Optional[int] = None,
+        distr_output: DistributionOutput = StudentTOutput(),
+        dropout_rate: float = 0.1,
+        lags_seq: Optional[List[int]] = None,
+        scaling: bool = True,
+        num_parallel_samples: int = 100,
+    ) -> None:
+        super().__init__()
+        self.num_layers = num_layers
+        self.num_cells = num_cells
+        self.cell_type = cell_type
+        self.context_length = context_length or prediction_length
+        self.prediction_length = prediction_length
+        self.dropout_rate = dropout_rate
+        self.num_feat_dynamic_real = num_feat_dynamic_real
+        self.num_feat_static_cat = num_feat_static_cat
+        self.num_feat_static_real = num_feat_static_real
+        self.cardinality = cardinality
+        self.embedding_dimension = (
+            embedding_dimension
+            if embedding_dimension is not None or self.cardinality is None
+            else [min(50, (cat + 1) // 2) for cat in self.cardinality]
+        )
+        self.scaling = scaling
+        self.lags_seq = lags_seq or get_lags_for_frequency(
+            freq_str=freq, lag_ub=self.context_length
+        )
+        self.num_parallel_samples = num_parallel_samples
+        self.history_length = self.context_length + max(self.lags_seq)
+        # for decoding the lags are shifted by one, at the first time-step
+        # of the decoder a lag of one corresponds to the last target value
+        self.shifted_lags = [l - 1 for l in self.lags_seq]
+        self.distr_output = distr_output
+        rnn = {"LSTM": nn.LSTM, "GRU": nn.GRU}[self.cell_type]
+        self.rnn = rnn(
+            input_size=self._rnn_input_size,
+            hidden_size=num_cells,
+            num_layers=num_layers,
+            dropout=dropout_rate,
+            batch_first=True,
+        )
+        self.target_shape = distr_output.event_shape
+        self.proj_distr_args = distr_output.get_args_proj(num_cells)
+        if self.cardinality is not None:
+            self.embedder = FeatureEmbedder(
+                cardinalities=self.cardinality,
+                embedding_dims=self.embedding_dimension,
+            )
+        else:
+            self.embedder = None
```

Comment:
I think let's make `cardinality` compulsory in the network for now. I think this is a leftover from when I was trying to make inputs to `forward` optional: eventually that would be nice to have, so that if one sets `num_feat_xxx_yyy = 0` the coresponding input can be missing.